src/main/scala/com/fulcrumgenomics/fastq/FastqSource.scala

```diff
@@ -0,0 +1,113 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package com.fulcrumgenomics.fastq
+
+import java.io._
+
+import com.fulcrumgenomics.util.Io
+import dagr.commons.CommonsDef.{PathToFastq, yieldAndThen}
+
+/**
+  * Provides factory methods for creating a `FastqSource` from multiple types.
+  */
+object FastqSource {
+  /** Creates a new fastq source from a Reader. */
+  def apply(reader: Reader): FastqSource = new FastqSource(reader)
+
+  /** Creates a new fastq source from a File. */
+  def apply(file: File): FastqSource = apply(path=file.toPath)
+
+  /** Creates a new fastq source from a Path. */
+  def apply(path: PathToFastq): FastqSource = apply(Io.toReader(path))
+}
+
+
+/**
+  * Reads fastq records from any text based source via a reader. Ensures that lines come in
+  * the expected groupings of four lines with the correct headers, and that bases and qualities
+  * are of the same length.  The underlying reader is closed automatically when EOF is reached.
+  */
+class FastqSource private(in: Reader) extends Iterator[FastqRecord] with Closeable {
+  private val reader : BufferedReader = in match {
+    case in : BufferedReader => in
+    case in : Reader         => new BufferedReader(in)
+  }
+
+  private var nextRecord: Option[FastqRecord] = None
+  queueNextRecord()
+
+  /** True if calling `next()` will yield another record, false otherwise. */
+  override def hasNext: Boolean = this.nextRecord.isDefined
+
+  /** Returns the next record if available, or throws an exception if none is available. */
+  override def next(): FastqRecord =  yieldAndThen(nextRecord.get) { queueNextRecord() }
```

Comment:
`nextRecord.get` will throw a `NoSuchElementException` if it's `None`, so I don't really see the point.  There is a test that asserts calling `next` when there is no `next` throws the right exception, so we'll catch it if we accidentally change it at some point.